### PR TITLE
Fix the jobs tab in the beta dashboard and fill it with data from both "submission" jobs and "driver" jobs

### PR DESCRIPTION
**Pull request by alanwguo**
## Why are these changes needed?
- Fixes the jobs tab in the new dashboard. Previously it didn't load.
- Combines the old job concept, "driver jobs" and the new job submission conception into a single concept called "jobs". Jobs tab shows information about both jobs.

- Updates all job APIs: They now returns both submission jobs and driver jobs. They also contains additional data in the response including "id", "job_id", "submission_id", and "driver". They also accept either job_id or submission_id as input.

- Job ID is the same as the "ray core job id" concept. It is in the form of "0100000" and is the primary id to represent jobs.
- Submission ID is an ID that is generated for each ray job submission. It is in the form of "raysubmit_12345...". It is a secondary id that can be used if a client needs to provide a self-generated id. or if the job id doesn't exist (ex: if the submission job doesn't create a ray driver)

This PR has 2 deprecations
- The `submit_job` sdk now accepts a new kwarg `submission_id`. `job_id is deprecated.
- The `ray job submit` CLI now accepts `--submission-id`. `--job-id` is deprecated.

**This PR has 4 backwards incompatible changes:**
- list_jobs sdk now returns a list instead of a dictionary
- the `ray job list` CLI now prints a list instead of a dictionary
- The `/api/jobs` endpoint returns a list instead of a dictionary
- The `POST api/jobs` endpoint (submit job) now returns a json with `submission_id` field instead of `job_id`.

Screen Shot 2022-07-12 at 3 31 14 PM

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/job/common.py
Comment (alanwguo) [inline dashboard/modules/job/common.py:90]: Actually, I probably shouldn't include this in this dataclass since driver is only filled out on the read path and is not ever stored in the JobInfo KV store.

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [inline dashboard/modules/job/job_head.py:251]: this is how `snapshot_head` gets driver info. Let me know if there's a better way.

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:219]: We need await here too (otherwise, this can block the main thread which is bad for API server).

I believe await is not supported right now due to some tech debt. Can you add TODO here (to use await list_jobs())?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:244]: Can you follow https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#code-style for the docstring?

File: dashboard/modules/job/job_head.py
Comment (edoakes) [reply dashboard/modules/job/job_head.py:219]: should be straightforward, just need to use async internal_kv API I think

File: dashboard/modules/job/job_head.py
Comment (architkulkarni) [inline dashboard/modules/job/job_head.py:269]: I guess now we're using JobInfo to describe both "drivers" and "submission jobs".  Do we need to redefine some of the fields to adapt it to both?  For example `entrypoint` wasn't optional, because all submission jobs had entrypoints, but now perhaps it should be optional because drivers don't have them.

This would break backward compatibility though for consumers of the snapshot API, so maybe we shouldn't do this.

File: dashboard/modules/job/job_head.py
Comment (architkulkarni) [reply dashboard/modules/job/job_head.py:251]: It looks like this is how it's done for legacy jobs, but for submission jobs we use a different API `self._job_info_client.get_all_jobs().items()` which ultimately pulls from the internal KV.  So would the alternative way to unify them be to store driver info in the internal KV and use `job_info_client` API?  

Anyways, it seems like your approach is simpler, and is async which is another advantage (though like Sang said, we should probably convert the `job_info_client` and `job_manager` APIs to use the async internal_kv API if they're going to be used in the API server.)

File: dashboard/modules/job/common.py
Comment (architkulkarni) [reply dashboard/modules/job/common.py:90]: Ah good point, I guess the problem with this is that in the future someone might read a JobInfo from the internal_kv and see `driver: None` and incorrectly conclude that there's no driver?  On the other hand, maybe this is an implementation detail, and we should add it to the dataclass now in case in the future we decide to populate the `driver` field earlier.

Comment (alanwguo) [reply dashboard/modules/job/common.py:90]: Yeah, this is a code smell to me that I'm unnecessarily re-using the `JobInfo` class. 

I think it's probably safer to not add it here for now and that also would make it less confusing for developers who would expect it to be filled out but it never is.

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [reply dashboard/modules/job/job_head.py:219]: I'll add a TODO. I don't see an async internal_kv API though.

https://github.com/ray-project/ray/blob/master/python/ray/experimental/internal_kv.py

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [reply dashboard/modules/job/job_head.py:248]: last, thanks for catching.

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [reply dashboard/modules/job/job_head.py:269]: I think we will eventually make this change when we move over to pydantic.

File: dashboard/client/src/pages/job/index.tsx
Comment (alanwguo) [reply dashboard/client/src/pages/job/index.tsx:71]: This page was no longer working anyways, so I'm just removing functionality that isn't supported in the new jobs api.

**Comment by rkooo567** [comment]
Btw @alanwguo the policy is OSS is to add reviewer to the "assignee" section (I already did!)

File: dashboard/modules/job/job_head.py
Comment (architkulkarni) [reply dashboard/modules/job/job_head.py:225]: I think you need to use `GcsAioClient`, see https://github.com/ray-project/ray/pull/25788 for an example

**Comment by architkulkarni** [comment]
New changes and description look good.  Could you clarify this part? "In the future, once all ray submission APIS accept job ids as input, we should always use job id."  

I'll defer to @tchordia on backwards compatibility for consumers of the cluster snapshot

**Comment by alanwguo** [comment]
> New changes and description look good. Could you clarify this part? "In the future, once all ray submission APIS accept job ids as input, we should always use job id."
> 
> I'll defer to @tchordia on backwards compatibility for consumers of the cluster snapshot

I want to change it so "id" field will always == "job id". I will only do that after the submission apis can accept both job id or submission ids. That's because most people will assume the id field is the thing they can use as the path parameter because its common practice in REST.

**Comment by alanwguo** [comment]
@edoakes, @rkooo567, can you review this code? I don't want to work on fixing tests until I get an OK on the overall approach.

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:54]: This would be great. We only used dataclass to avoid adding pydantic as a dependency. Once it's added we should use it for all of these.

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:252]: should we use node_id additionally / instead of ip address?

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:265]: can we use an enum instead of raw string here?

File: dashboard/modules/job/common.py
Comment (edoakes) [inline dashboard/modules/job/common.py:280]: It's pretty confusing that this thing is described as "the id for a driver launch in ray" but it's called "**job**_id" :(

**Review by edoakes — COMMENTED** [review]
@alanwguo the implementation looks reasonable to me, though it'd be great to convert fetching the job info to use `async` (can do that later).

I'm still pretty dissatisfied with the ID situation to be honest, but maybe this is the best we can do. Why do we need both `id` and the others? Is there a case where it's important to have a primary ID? Or could we just have `submission_id` and `job_id`?

**Comment by alanwguo** [comment]
> @alanwguo the implementation looks reasonable to me, though it'd be great to convert fetching the job info to use `async` (can do that later).
> 
> I'm still pretty dissatisfied with the ID situation to be honest, but maybe this is the best we can do. Why do we need both `id` and the others? Is there a case where it's important to have a primary ID? Or could we just have `submission_id` and `job_id`?

I just think a primary id field is useful for folks who are used to REST semantics. Otherwise, it's not obvious which ID to pass in when calling the `/jobs/{id}` endpoint. If that endpoint accepts either type of ID, maybe that's okay...

File: dashboard/modules/job/common.py
Comment (alanwguo) [reply dashboard/modules/job/common.py:252]: i can add the node_id as a separate field but the ip address is actually useful for the oss dashboard UI because it's required to link to the logs page for specific nodes.

File: dashboard/modules/job/common.py
Comment (alanwguo) [reply dashboard/modules/job/common.py:54]: I'll tackle this as a follow-up since this is a much riskier change because it requires changes to the actual jobs source of truth data storage.

File: dashboard/modules/job/common.py
Comment (alanwguo) [reply dashboard/modules/job/common.py:252]: I decided not to add `node_id` for now. It's not easily accessible in the `jobs_table_entry protobuf`.

File: dashboard/client/src/pages/job/index.tsx
Comment (rkooo567) [inline dashboard/client/src/pages/job/index.tsx:29]: can we keep the pid? it is useful to get the corresponding driver logs (cpp)

Comment (rkooo567) [reply dashboard/client/src/pages/job/index.tsx:29]: maybe it could be at the end of the col

File: dashboard/client/src/pages/job/index.tsx
Comment (rkooo567) [inline dashboard/client/src/pages/job/index.tsx:139]: Can we also add a running_time column?  (could be done in the future PRs though). But I think that'd be pretty useful

File: dashboard/modules/job/common.py
Comment (rkooo567) [inline dashboard/modules/job/common.py:8]: should remove it?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:278]: Specify the return type?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:281]: Why don't we use a named tuple as a return value?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [reply dashboard/modules/job/job_head.py:287]: I think this form is okay. We should just use state APIs after we migrate all these in the future

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [reply dashboard/modules/job/job_head.py:281]: hmm actually why do we need to distinguish them at all? maybe we can return both types of jobs to the same data structure?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [reply dashboard/modules/job/job_head.py:255]: can we wrap this to thread executor in this case? I'd like to avoid having blocking call in a periodically called API (it's called from dashboard)

File: python/ray/scripts/scripts.py
Comment (rkooo567) [inline python/ray/scripts/scripts.py:2544]: why do we do this specifically for this command?

File: python/setup.py
Comment (rkooo567) [inline python/setup.py:223]: remove it since we don't do migration in this PR?

**Review by rkooo567 — COMMENTED** [review]
The frontend review is not well done btw (I tried though..)

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [reply dashboard/modules/job/job_head.py:281]: one is keyed by job_id and the other is keyed by submission_id for easier lookup.

File: python/ray/scripts/scripts.py
Comment (alanwguo) [reply python/ray/scripts/scripts.py:2544]: jobs cli depends on pydantic which is only included in with ray[default]. This is similar to serve cli below this line.

File: python/setup.py
Comment (alanwguo) [reply python/setup.py:223]: I am using pydantic in this PR.

File: dashboard/modules/job/common.py
Comment (alanwguo) [reply dashboard/modules/job/common.py:8]: it's used in this file.

File: dashboard/client/src/pages/job/index.tsx
Comment (alanwguo) [reply dashboard/client/src/pages/job/index.tsx:140]: I'll do this in a follow-up..

**Comment by alanwguo** [comment]
@edoakes @ericl @architkulkarni @rkooo567 , 

Wanted to call out some all the backwards-incompatible changes:

This PR has 2 deprecations
- The `submit_job` sdk now accepts a new kwarg `submission_id`. `job_id is deprecated.
- The `ray job submit` CLI now accepts `--submission-id`. `--job-id` is deprecated.

**This PR has 4 backwards incompatible changes:**
- list_jobs sdk now returns a list instead of a dictionary
- the `ray job list` CLI now prints a list instead of a dictionary
- The `/api/jobs` endpoint returns a list instead of a dictionary
- The `POST api/jobs` endpoint (submit job) now returns a json with `submission_id` field instead of `job_id`.

**Comment by rkooo567** [comment]
cc @edoakes for the backward incompatible change.

We'd like to merge this by the branch cut right?

File: dashboard/modules/job/tests/test_cli_integration.py
Comment (rkooo567) [inline dashboard/modules/job/tests/test_cli_integration.py:22]: what is this for btw?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:99]: Can you double check if `get_job_info ` is thread-safe?

File: dashboard/modules/job/job_head.py
Comment (rkooo567) [inline dashboard/modules/job/job_head.py:260]: Can you double check and confirm list_jobs is thread-safe? (I guess yes?)

**Review by rkooo567 — APPROVED** [review]

File: dashboard/modules/job/tests/test_cli_integration.py
Comment (alanwguo) [reply dashboard/modules/job/tests/test_cli_integration.py:22]: This makes the tests less leaky in the case of an exception being thrown in the test code. If there is an exception during the test, previously, the env vars were not being reset.

File: dashboard/modules/job/job_head.py
Comment (edoakes) [reply dashboard/modules/job/job_head.py:99]: @rkooo567 this is essentially directly an internal_kv call, so it should be thread safe.

In general, I'd much prefer to refactor the job manager to make it `asyncio` natively instead of using `run_in_executor` though because it's easy to introduce unexpected behaviors with this. @alanwguo is there anything blocking this? I would like to do it as an immediate follow-up if not in this PR.

**Review by edoakes — COMMENTED** [review]
Looks good, only concern is about using `run_in_executor`. Not blocking merge given the timeline for 2.0 branch cut but please make sure to follow up on this.

File: dashboard/modules/job/job_head.py
Comment (alanwguo) [reply dashboard/modules/job/job_head.py:99]: I'll do it as a follow-up. No blockers, just want to avoid refactoring a bunch of code in this PR.

**Comment by rkooo567** [comment]
> In general, I'd much prefer to refactor the job manager to make it asyncio natively instead of using run_in_executor though because it's easy to introduce unexpected behaviors with this.

Yeah we should definitely do that asap. I think it should be possible right now.

**Comment by rkooo567** [comment]
@architkulkarni Is there a way to verify this doesn't break the prod job submission asap?

**Comment by architkulkarni** [comment]
I checked the code to make sure it wasn't using any APIs that were broken.  I'll manually test it now using the generated wheel for this commit.  I don't think there's another way to verify it.

File: dashboard/modules/job/job_head.py
Comment (architkulkarni) [inline dashboard/modules/job/job_head.py:114]: Sorry for not seeing this earlier, it looks like we can use this endpoint in the future to make breaking API changes smoother for third-party job managers that use our API.

Comment (alanwguo) [reply dashboard/modules/job/job_head.py:114]: ill bump this

File: dashboard/client/src/pages/job/index.tsx
Comment (rkooo567) [reply dashboard/client/src/pages/job/index.tsx:140]: btw ping on this as a follow up

Comment (rkooo567) [reply dashboard/client/src/pages/job/index.tsx:140]: Btw, I think we can do this after 2.0.

**Pull request merged**
